File: frontend/src/stores/aiChats.js
import { defineStore } from 'pinia'
import api from '../services/api'

export const useAIChatsStore = defineStore('aiChats', {
  state: () => ({
    chats: [],
    currentChat: null,
    messages: [],
    loading: false,
    sending: false,
    error: null,
  }),

  getters: {
    sortedChats: (state) => [...state.chats].sort(
      (a, b) => new Date(b.updated_at) - new Date(a.updated_at)
    ),
  },

  actions: {
    async fetchChats() {
      this.loading = true
      this.error = null
      try {
        const res = await api.get('/ai/chats')
        this.chats = res.data
        return res.data
      } catch (err) {
        this.error = err.response?.data?.detail || 'Не удалось загрузить AI-чаты'
        throw err
      } finally {
        this.loading = false
      }
    },

    async createChat(title = null) {
      this.loading = true
      this.error = null
      try {
        const res = await api.post('/ai/chats', { title })
        this._upsertChat(res.data)
        this.currentChat = res.data
        this.messages = []
        return res.data
      } catch (err) {
        this.error = err.response?.data?.detail || 'Не удалось создать AI-чат'
        throw err
      } finally {
        this.loading = false
      }
    },

    async fetchChat(chatId) {
      this.loading = true
      this.error = null
      try {
        const res = await api.get(`/ai/chats/${chatId}`)
        const { messages = [], ...chat } = res.data
        this.currentChat = chat
        this.messages = messages
        this._upsertChat(chat)
        return res.data
      } catch (err) {
        this.error = err.response?.data?.detail || 'Не удалось открыть AI-чат'
        throw err
      } finally {
        this.loading = false
      }
    },

    async sendMessage(chatId, content) {
      this.sending = true
      this.error = null
      try {
        const res = await api.post(`/ai/chats/${chatId}/messages`, { content })
        this._appendMessage(res.data.user_message)
        this._appendMessage(res.data.assistant_message)
        this._upsertChat(res.data.chat)
        this.currentChat = res.data.chat
        if (res.data.error) this.error = res.data.error
        return res.data
      } catch (err) {
        this.error = err.response?.data?.detail || 'AI сейчас недоступен'
        throw err
      } finally {
        this.sending = false
      }
    },

    async analyzeReport(reportId) {
      this.loading = true
      this.error = null
      try {
        const res = await api.post(`/ai/reports/${reportId}/analyze`)
        this._applyContextResponse(res.data)
        return res.data.chat
      } catch (err) {
        this.error = err.response?.data?.detail || 'Не удалось отправить отчет в AI'
        throw err
      } finally {
        this.loading = false
      }
    },

    async planGoal(goalId) {
      this.loading = true
      this.error = null
      try {
        const res = await api.post(`/ai/goals/${goalId}/plan`)
        this._applyContextResponse(res.data)
        return res.data.chat
      } catch (err) {
        this.error = err.response?.data?.detail || 'Не удалось составить план с AI'
        throw err
      } finally {
        this.loading = false
      }
    },

    _applyContextResponse(data) {
      this.currentChat = data.chat
      this.messages = [data.user_message, data.assistant_message]
      this._upsertChat(data.chat)
      if (data.error) this.error = data.error
    },

    _appendMessage(message) {
      const existingIndex = this.messages.findIndex(item => item.id === message.id)
      if (existingIndex === -1) {
        this.messages.push(message)
      } else {
        this.messages[existingIndex] = message
      }
    },

    async renameChat(chatId, title) {
      this.loading = true
      this.error = null
      try {
        const res = await api.patch(`/ai/chats/${chatId}`, { title })
        this._upsertChat(res.data)
        if (this.currentChat?.id === chatId) this.currentChat = res.data
        return res.data
      } catch (err) {
        this.error = err.response?.data?.detail || 'Не удалось переименовать чат'
        throw err
      } finally {
        this.loading = false
      }
    },

    async deleteChat(chatId) {
      this.loading = true
      this.error = null
      try {
        await api.delete(`/ai/chats/${chatId}`)
        this.chats = this.chats.filter(c => c.id !== chatId)
        if (this.currentChat?.id === chatId) {
          this.currentChat = null
          this.messages = []
        }
      } catch (err) {
        this.error = err.response?.data?.detail || 'Не удалось удалить чат'
        throw err
      } finally {
        this.loading = false
      }
    },

    _upsertChat(chat) {
      const index = this.chats.findIndex(item => item.id === chat.id)
      if (index === -1) {
        this.chats.unshift(chat)
      } else {
        this.chats[index] = chat
      }
    },
  },
})

File: backend/app/api/ai_chats.py
from fastapi import APIRouter, Depends, HTTPException, Query

from app.core.dependencies import get_current_user
from app.core.middleware import check_ai_rate_limit
from app.models.ai_chat import AIChatCreate, AIChatUpdate, AIMessageCreate
from app.services import ai_chat_service

router = APIRouter(prefix="/api/ai", tags=["AI Chat"])


async def _require_ai_quota(current_user: dict) -> None:
    """Raise 429 if the user has exceeded their hourly AI call quota."""
    if not await check_ai_rate_limit(current_user["id"]):
        raise HTTPException(
            status_code=429,
            detail="Превышен лимит AI-запросов (20 в час). Попробуйте позже.",
        )


@router.get("/chats")
async def list_ai_chats(
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    current_user: dict = Depends(get_current_user),
):
    return await ai_chat_service.get_user_chats(current_user["id"], skip, limit)


@router.post("/chats")
async def create_ai_chat(
    data: AIChatCreate,
    current_user: dict = Depends(get_current_user),
):
    return await ai_chat_service.create_chat(current_user["id"], data)


@router.get("/chats/{chat_id}")
async def get_ai_chat(
    chat_id: str,
    current_user: dict = Depends(get_current_user),
):
    return await ai_chat_service.get_chat_detail(chat_id, current_user["id"])


@router.patch("/chats/{chat_id}")
async def rename_ai_chat(
    chat_id: str,
    data: AIChatUpdate,
    current_user: dict = Depends(get_current_user),
):
    return await ai_chat_service.rename_chat(chat_id, current_user["id"], data.title)


@router.delete("/chats/{chat_id}")
async def delete_ai_chat(
    chat_id: str,
    current_user: dict = Depends(get_current_user),
):
    await ai_chat_service.delete_chat(chat_id, current_user["id"])
    return {"detail": "Chat deleted"}


@router.post("/chats/{chat_id}/messages")
async def send_ai_message(
    chat_id: str,
    data: AIMessageCreate,
    current_user: dict = Depends(get_current_user),
):
    await _require_ai_quota(current_user)
    return await ai_chat_service.send_message(chat_id, current_user["id"], data.content)


@router.post("/reports/{report_id}/analyze")
async def analyze_report_with_ai(
    report_id: str,
    current_user: dict = Depends(get_current_user),
):
    await _require_ai_quota(current_user)
    return await ai_chat_service.analyze_report(report_id, current_user)


@router.post("/goals/{goal_id}/plan")
async def plan_goal_with_ai(
    goal_id: str,
    current_user: dict = Depends(get_current_user),
):
    await _require_ai_quota(current_user)
    return await ai_chat_service.plan_goal(goal_id, current_user)
